app/routes/auth.js

```diff
@@ -0,0 +1,180 @@
+const express = require('express')
+const router = express.Router()
+const crypto = require('crypto')
+const { PrismaClient } = require('@prisma/client');
+const { PrismaClientKnownRequestError } = require('@prisma/client/runtime/library');
+const prisma = new PrismaClient();
+const passport = require('passport')
+const LocalStrategy = require('passport-local')
+const passwordValidator = require('password-validator')
+const emailValidator = require('email-validator')
+
+// Configuring passport strategy.
+passport.use(new LocalStrategy(async function verify(email, password, done) {
+    let user = await prisma.user.findUnique({
+        where: {
+            email: email
+        }
+    })
+    if (!user) return done(null, false, { message: 'Incorrect login credentials.' })
+
+    crypto.pbkdf2(password, user.salt, 310000, 32, 'sha256', async (err, hashedPassword) => {
+        if (err) return done(err)
+        if (!crypto.timingSafeEqual(user.hashed_password, hashedPassword)) {
+            return done(null, false, { message: 'Incorrect login credentials.' });
+        }
+
+        return done(null, user)
+    })
+}))
+
+// Serialization Functions
+passport.serializeUser(function (user, done) {
+    process.nextTick(function () {
+        done(null, { id: user.id, email: user.email, fname: user.fname, lname: user.lname });
+    })
+});
+
+passport.deserializeUser(function (user, done) {
+    process.nextTick(function () {
+        return done(null, user);
+    })
+});
+
+// Middleware that redirects to login page if user is not logged in.
+function isLoggedIn(req, res, next) {
+    req.user ? next() : res.redirect('/login')
+}
+
+// Route for log in page.
+router.get('/login', (req, res, next) => {
+    let error
+    if (req.query.error && req.session && req.session.messages) {
+        error = req.session.messages
+    }
```

Comment:
Session messages are very cool but appear to have the downside that if I log in multiple times in a row with incorrect credentials (or if I navigate away and come back and try again), the error box fills with the same error message.

See:
![Example](https://github.com/LukarioMC/COMP-4800-37/assets/15253958/2fafe894-78a8-4b62-bde3-a921dd2a64a5)

Reply:
Changed it to pop the last message so that only 1 messsage is shown and it is removed.